types/src/error.rs

```diff
@@ -104,6 +105,36 @@ impl<'a> PartialEq for ErrorObject<'a> {
 	}
 }
 
+/// JSON-RPC error object.
+#[derive(Debug, Deserialize, Serialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct ErrorObjectOwned {
+	/// Code
+	pub code: ErrorCode,
+	/// Message
+	pub message: String,
+	/// Optional data
+	#[serde(skip_serializing_if = "Option::is_none")]
+	pub data: Option<Box<RawValue>>,
+}
```

Comment:
This may be a bad idea, but I wonder whether we can unify `ErrorObject` and `ErrorObjectOwned` with some `Cow`s, and pthen erhaps provide `borrowed(code: i32, message: &str, data: Option<&'a RawValue>) -> ErrorObject<'a>` and `owned(code: i32, message: String, data: Option<Box<RawValue>>) -> ErrorObject<'static>` constructors for it?

Reply:
(I'll have a quick play and see whether it's actually possible)

Reply:
Ok, so how about this (ignore the `ErrorObject2` naming; just playing around in the existing codebase so I needed a name that didn't overlap!):

```rust
use std::borrow::{ Cow as StdCow };
use std::borrow::Borrow;

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(deny_unknown_fields)]
pub struct ErrorObject2<'a> {
	/// Code
	code: ErrorCode,
	/// Message
	message: StdCow<'a, str>,
	/// Optional data
	#[serde(skip_serializing_if = "Option::is_none")]
	data: Option<StdCow<'a, RawValue>>,
}

pub type ErrorObject2Owned = ErrorObject2<'static>;

impl ErrorObject2Owned {
	/// Return the error code
	pub fn code(&self) -> i32 {
		self.code.code()
	}

	/// Return the message
	pub fn message(&self) -> &str {
		self.message.borrow()
	}

	/// Return the data associated with this error, if any
	pub fn data(&self) -> Option<&RawValue> {
		self.data.as_ref().map(|d| d.borrow())
	}

	/// Create a new `ErrorObjectOwned` with optional data.
	pub fn owned(code: i32, message: impl Into<String>, data: Option<impl Serialize>) -> Self {
		let data = data.and_then(|d| serde_json::value::to_raw_value(&d).ok());
		ErrorObject2 {
			code: code.into(),
			message: message.into().into(),
			data: data.map(|d| StdCow::Owned(d))
		}
	}

	/// Create a new `ErrorObject` with optional data.
	pub fn borrowed<'a>(code: i32, message: &'a str, data: Option<&'a RawValue>) -> ErrorObject2<'a> {
		ErrorObject2 {
			code: code.into(),
			message: StdCow::Borrowed(message),
			data: data.map(|d| StdCow::Borrowed(d))
		}
	}

	/// Get the borrowed variant [`ErrorObject`].
	pub fn into_owned(self) -> ErrorObject2<'static> {
		ErrorObject2 {
			code: self.code,
			message: StdCow::Owned(self.message.into_owned()),
			data: self.data.map(|d| StdCow::Owned(d.into_owned()))
		}
	}

	/// Get the borrowed variant [`ErrorObject`].
	pub fn borrow<'a>(&'a self) -> ErrorObject2<'a> {
		ErrorObject2 {
			code: self.code,
			message: StdCow::Borrowed(self.message.borrow()),
			data: self.data.as_ref().map(|d| StdCow::Borrowed(d.borrow()))
		}
	}
}
```

Reply:
(I made both owned and borrowed functions accept `Option<data>` for consistency; with the separate borrowed and owned variants, one accepts Option and one does not)

Reply:
looks good to me the reason why it's like is currently is because `beef::Cow` doesn't work with RawValue.

Reply:
That makes sense; `beef::Cow` will only ever support certain types I guess, so we should feel free to use the std `Cow` where useful! (perhaps I could have used the beef variant for strings above; I can't recall!)